Extend default mapping

iuljan opened the pull request:
Close #26

Commit:
Add test for default parameters

Commit:
Extend global_paramter_map

Commit:
Add  parameter for extended parameter

Commit:
Add adapter class name to mapper

iuljan commented on data_adapter_oemof/mappings/GLOBAL_PARAMETER_MAP.yaml:
```diff
@@ -1,8 +1,13 @@
+  #overnight_cost: sedos-overnight_cost
+  #lifetime: sedos-lifetime
+  #wacc: sedos-wacc
+WindOnshore:
```
@henhuy Should this be the classname e.g. `VolatileAdapter` or more specific like right now? In this case, the `technology` information needs to be passed by your adapter or come from somewhere else...

henhuy replied:
For now, I think this should be the class name, aka the str you get when you do `cls.__name__`, which is `VolatileAdapter` in the upper case.

iuljan commented on data_adapter_oemof/mappings.py:
```diff
@@ -12,11 +12,12 @@ class MappingError(Exception):
             mapping = GLOBAL_PARAMETER_MAP
         if bus_map is None:
             bus_map = BUS_NAME_MAP
+        self.adapter_clsname = adapter_clsname
```
I added this as discussed but now I am not sure what its purpose is? Is this to be used for distinguishing the second level of global_parameter_map?

henhuy replied:
Instead, you could also pass whole adapter to the mapper, like in adapter:
```
# in Adapter:
mapper = Mapper(cls, ...)

# in Mapper:
def __init__(self, adapter, ...):
    self.adapter = adapter
    ...
```
And read name later via `self.adapter.__class__.__name__`

Commit:
Pass adapter to mapper instead of name only

Commit:
Use adapter-name for stepwise parameter mapping

Commit:
Correct test

Commit:
Pylint

henhuy commented on tests/test_mapping.py:
```diff
@@ -20,13 +23,41 @@ def test_get_with_mapping():
     assert mapped == expected
 
 
 def test_get_defaults():
```
This tests includes too many setups and assumptions.
I would instead only test outputs of Mapper class with custom mapping, like:
```
mapping= {"DEFAULT": {...}, "Adapter": {...}}
mapper = Mapper(Adapter, mapping=mapping)
assert ...
```
This makes the test independent from changes in adapter class aka `parametrize_dataclass` or `as_dict`...

iuljan replied:
Don't really understand your comment. Isn't that what's `test_get_with_mapping` is doing? I guess you talk about `test_get_defaults` which tests much more.

henhuy replied:
yes, i meant `test_get_defaults` - don't know why I selected the wrong lines...
So my comment is, that `test_get_defaults` does too much (too much dependencies).
Unit testing is about testing small independent units

iuljan replied:
removed it

henhuy commented on data_adapter_oemof/mappings.py:
```diff
@@ -12,18 +12,25 @@ class MappingError(Exception):
         if bus_map is None:
             bus_map = BUS_NAME_MAP
+        self.adapter = adapter
+        self.adapter_name = adapter.__name__
```
would only store adapter

Commit:
Only store adapter

Commit:
Merge branch 'dev' into extend-default-mapping

Commit:
Adapt test to timeseries

Commit:
Adapt tests to timeseries-feature

Commit:
Fix bug to still pass custom mapping

Commit:
Black

Commit:
Add test for sequences

Commit:
Pass data input as DataFrame

Commit:
Adjust sequence test

iuljan commented:
@henhuy Should the default values from the tabular facade be used if no values can be found?

Commit:
Fix sequence mapping

Mapped_key is alywas a string, as data is passed to
parametrize_dataclass as dictionary. String cant be
converted to a list. This was probably coded under the
assumption that data is passed as a DataFrame.

Commit:
Use mapped_key for last default mapping option.

Commit:
Fix mapping tests and remove to big test

henhuy commented:
> @henhuy Should the default values from the tabular facade be used if no values can be found?

Yes, definitely!

henhuy reviewed (approved):
Looks good to me!
Nice work.

FelixMau committed:
Merge branch 'dev' into extend-default-mapping

FelixMau commented:
Merged because I did not see you already fixed faulty test in `test_mappings`

henhuy committed:
Return None if mapping fails

Commit:
Extend comments

Commit:
Include 'type' in parameter defaults

Commit:
Fix typo

Commit:
Split get() into mapping and data function

Commit:
Replace TYPE_MAP with FACADE_ADAPTERS

FACADE_ADAPTERS is an auto-generated dict containing
all defined adapters. This is necesary to connect the
the new PROCESS_TYPE_MAP.yaml with the actual adapter
classes.

Commit:
Directly map processes to adapters within PROCESS_TYPE_MAP.yaml

Adapters are directly defined in this file. FACADE_ADAPTERS is
used to link the actual adapters.

Commit:
Implement additional parameter mapping hierachy

The key mapping now takes place with an additional
level. First the process specific mapping is checked,
if not available, the facade-specific mapping is used
and then the defaults.

Commit:
Black

Commit:
Adapt tests

henhuy committed:
Refactor map_key to early-return-style
